singularity-eos/eos/eos_powermg.hpp

```diff
@@ -0,0 +1,440 @@
+               AllHugTemperatureFromDensity(rho);
+  if (value < 0.0) {
+#ifndef NDEBUG
+    PORTABLE_WARN("Negative temperature");
+#endif // NDEBUG
+    value = 1.e-12;
+  }
+  return value;
+}
+PORTABLE_INLINE_FUNCTION Real PowerMG::PressureFromDensityInternalEnergy(
+    const Real rho, const Real sie, Real *lambda) const {
+  Real value = AllHugPressureFromDensity(rho) +
+               _rho0 * _G0 * (sie - AllHugInternalEnergyFromDensity(rho));
+  return value;
+}
+PORTABLE_INLINE_FUNCTION Real PowerMG::MinInternalEnergyFromDensity(const Real rho,
+                                                                    Real *lambda) const {
+  MinInternalEnergyIsNotEnabled("PowerMG");
+  return 0.0;
+}
+PORTABLE_INLINE_FUNCTION Real PowerMG::EntropyFromDensityInternalEnergy(
+    const Real rho, const Real sie, Real *lambda) const {
+  Real eta = 1.0 - robust::ratio(_rho0, rho);
+  Real value = std::log(TemperatureFromDensityInternalEnergy(rho, sie) / _T0);
+  value = _S0 - _G0 * _Cv0 * eta + _Cv0 * value;
+  if (value < 0.0) {
+#ifndef NDEBUG
+    PORTABLE_WARN("Negative entropy");
+#endif // NDEBUG
+    value = 1.e-12;
+  }
+  return value;
+}
+// AEM: Give error since function is not well defined
+PORTABLE_INLINE_FUNCTION void
+PowerMG::DensityEnergyFromPressureTemperature(const Real press, const Real temp,
+                                              Real *lambda, Real &rho, Real &sie) const {
+  EOS_ERROR("PowerMG::DensityEnergyFromPressureTemperature: "
+            "Not implemented.\n");
+}
+// AEM: We should add entropy and Gruneissen parameters here so that it is complete
+// If we add also alpha and BT, those should also be in here.
+PORTABLE_INLINE_FUNCTION void PowerMG::FillEos(Real &rho, Real &temp, Real &sie,
+                                               Real &press, Real &cv, Real &bmod,
+                                               const unsigned long output,
+                                               Real *lambda) const {
+  const unsigned long input = ~output; // everything that is not output is input
+  if (thermalqs::density & output) {
+    EOS_ERROR("PowerMG FillEos: Density is required input.\n");
+  } else if (thermalqs::temperature & output &&
+             thermalqs::specific_internal_energy & output) {
+    EOS_ERROR("PowerMG FillEos: Density and Internal Energy or Density and Temperature "
+              "are required input parameters.\n");
+  }
+  if (thermalqs::temperature & input) {
+    sie = InternalEnergyFromDensityTemperature(rho, temp);
+  }
+  if (output & thermalqs::temperature)
+    temp = TemperatureFromDensityInternalEnergy(rho, sie);
+  if (output & thermalqs::specific_internal_energy) sie = sie;
+  if (output & thermalqs::pressure) press = PressureFromDensityInternalEnergy(rho, sie);
+  if (output & thermalqs::specific_heat)
+    cv = SpecificHeatFromDensityInternalEnergy(rho, sie);
+  if (output & thermalqs::bulk_modulus)
+    bmod = BulkModulusFromDensityInternalEnergy(rho, sie);
+}
```

Comment:
In Vinet, if I recall, you computed the free energy and then got everything from that, so the `FillEos` model could just compute the sum over coefficients once for all outputs. Is there a similar optimization you could do here? Where, e.g., you compute the sum over the coefficients once, and then get everything from that?

Reply:
OK. You did not like that ;).
I did not even think about doing it in that way. I basically followed how Robinson wrote it in the report. (referenced in the documentation). You want me to look into it?

Reply:
I didn't like it in the smaller API calls (like pressure from density temperature) because computing everything for that call implies wasted work. However, it's a win in `FillEos` because you compute things once and then leverage them for all the return values `FillEos` can provide. That's actually why `FillEos` exists---so one can optimize like that if desired.

My thinking was just that math operations, e.g., summing over the Taylor series could be reused in `FIllEos` for a performance gain. But if it's not easy to do, don't bother now. It's not worth your time.

Reply:
Done